pkg/cmd/cli/cmd/observe/observe.go

```diff
@@ -0,0 +1,936 @@
+	%[1]s observe services -a '{ .spec.clusterIP }' --each register_dns.sh
+`
+)
+
+// NewCmdObserve creates the observe command.
+func NewCmdObserve(fullName string, f *clientcmd.Factory, out, errOut io.Writer) *cobra.Command {
+	options := &ObserveOptions{
+		baseCommandName: fullName,
+		retryCount:      2,
+		templateType:    "jsonpath",
+		maximumErrors:   20,
+		listenAddr:      ":11251",
+	}
+
+	cmd := &cobra.Command{
+		Use:     "observe",
+		Short:   "Observe changes to resources and invoke integrations",
+		Long:    fmt.Sprintf(observeLong, fullName),
+		Example: fmt.Sprintf(observeExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := options.Complete(f, args, out, errOut); err != nil {
+				cmdutil.CheckErr(err)
+			}
+
+			if err := options.Validate(args); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(cmd, err.Error()))
+			}
+
+			if err := options.Run(); err != nil {
+				cmdutil.CheckErr(err)
+			}
+		},
+	}
+
+	// flags controlling what to select
+	cmd.Flags().BoolVar(&options.allNamespaces, "all-namespaces", false, "If present, list the requested object(s) across all projects. Project in current context is ignored.")
+
+	// the commands to execute on changes
+	cmd.Flags().VarP(&options.eachCommand, "each", "e", "A command to run for each change. Specify mulitple times to add arguments.")
+
+	// to perform deletion observehronization
+	cmd.Flags().Var(&options.nameSyncCommand, "names", "A command that will list all of the currently known names, optional. Specify multiple times to add arguments. Use to get notifications when objects are deleted.")
+	cmd.Flags().BoolVar(&options.includeDeletes, "include-deletes", false, "If true, the --each command will be called for deleted objects. Some deleted objects will not have generated arguments beyond namespace and name.")
+
+	// add additional arguments / info to the server
+	cmd.Flags().StringVar(&options.templateType, "use", options.templateType, "Format for template arguments. jsonpath and gotemplate are supported.")
+	cmd.Flags().BoolVar(&options.strictTemplates, "strict-templates", false, "If true return an error on any field or map key that is not missing in a template.")
+	cmd.Flags().VarP(&options.templates, "argument", "a", "Template for the arguments to be passed to each command in the format defined by --use.")
+	cmd.Flags().BoolVar(&options.includeType, "include-type-argument", false, "If true, include the change type ('Sync', 'Updated', 'Deleted', 'Added') as the first argument to the --each command.")
+	cmd.Flags().StringVar(&options.objectEnvVar, "env-var", "", "The name of an env var to serialize the object to when calling --each, optional.")
+
+	// control retries of individual commands
+	cmd.Flags().IntVar(&options.maximumErrors, "maximum-errors", options.maximumErrors, "Exit after this many errors have been detected with. May be set to -1 for no maximum.")
```

Comment:
I don't think so.  I think you're in a pod, and if you hit enough errors
you restart.  There is no sliding window that will work for all users.  We
could do % for sync but it wouldn't have any meaning after sync.  In
practice, your scripts should never error.  If you're erroring, you want to
restart.

On Tue, Aug 30, 2016 at 2:49 PM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/observe/observe.go
> <https://github.com/openshift/origin/pull/4196#discussion_r76856229>:
>
> > +	// the commands to execute on changes
> > +	cmd.Flags().VarP(&options.eachCommand, "each", "e", "A command to run for each change. Specify mulitple times to add arguments.")
> > +
> > +	// to perform deletion observehronization
> > +	cmd.Flags().Var(&options.nameSyncCommand, "names", "A command that will list all of the currently known names, optional. Specify multiple times to add arguments. Use to get notifications when objects are deleted.")
> > +	cmd.Flags().BoolVar(&options.includeDeletes, "include-deletes", false, "If true, the --each command will be called for deleted objects. Some deleted objects will not have generated arguments beyond namespace and name.")
> > +
> > +	// add additional arguments / info to the server
> > +	cmd.Flags().StringVar(&options.templateType, "use", options.templateType, "Format for template arguments. jsonpath and gotemplate are supported.")
> > +	cmd.Flags().BoolVar(&options.strictTemplates, "strict-templates", false, "If true return an error on any field or map key that is not missing in a template.")
> > +	cmd.Flags().VarP(&options.templates, "argument", "a", "Template for the arguments to be passed to each command in the format defined by --use.")
> > +	cmd.Flags().BoolVar(&options.includeType, "include-type-argument", false, "If true, include the change type ('Sync', 'Updated', 'Deleted', 'Added') as the first argument to the --each command.")
> > +	cmd.Flags().StringVar(&options.objectEnvVar, "env-var", "", "The name of an env var to serialize the object to when calling --each, optional.")
> > +
> > +	// control retries of individual commands
> > +	cmd.Flags().IntVar(&options.maximumErrors, "maximum-errors", options.maximumErrors, "Exit after this many errors have been detected with. May be set to -1 for no maximum.")
>
> I really don't like this. It's almost never what you want. You usually
> want a failure percentage based on a sliding window of attempts plus a
> min-attempts number.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/4196/files/69ab95173b3a274314ae59c42690a33690f9f69e#r76856229>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p5FUhJT4Prwank7WGof-J5gehtqjks5qlHshgaJpZM4Fsaph>
> .
>